test/parallel/test-stream-finished.js

```diff
@@ -396,3 +403,76 @@ testClosed((opts) => new Writable({ write() {}, ...opts }));
   r.destroyed = true;
   r.push(null);
 }
+
+{
+  // Regression https://github.com/nodejs/node/issues/33130
+
+  const response = new PassThrough();
+  response.write('chunk 1');
+  response.write('chunk 2');
+  response.write('chunk 3');
+  response.end();
+
+  class HelloWorld extends Duplex {
+    constructor(response) {
+      super();
+
+      this.response = response;
+      this.readMore = false;
+
+      response.once('end', () => {
+        this.push(null);
+      });
+
+      response.on('readable', () => {
+        if (this.readMore) {
+          this._read();
+        }
+      });
+    }
+
+    _read() {
+      const { response } = this;
+
+      this.readMore = true;
+
+      if (response.readableLength) {
+        this.readMore = false;
+      }
+
+      let data;
+      while ((data = response.read()) !== null) {
+        this.push(data);
+      }
+    }
+  }
+
+  const instance = new HelloWorld(response);
+  instance.setEncoding('utf8');
+  instance.end();
+
+  instance.on('finish', common.mustCall(async () => {
+    assert.strictEqual(instance.writableFinished, true);
+
+    let res = '';
+    for await (const data of instance) {
+      res += data;
+    }
+
+    assert.strictEqual(res, 'chunk 1chunk 2chunk 3');
+  }));
```

Comment:
I thinkhis still needs a `.then(common.mustCall())`, otherwise we don’t know that the async function ever finishes here

Reply:
fixed